### ISSUE TITLE
Color emoji in window titles cause taffybar to double in size

### ISSUE THREAD

**Issue by benley** [taffybar/taffybar#413]
**Describe the bug**
My taffybar is normally 55 pixels tall, and contains a window list widget, which usually works fine.  If a X11 window title happens to contain certain emoji characters, taffybar abruptly doubles its height to display the super high-res emoji glyphs.

**To Reproduce**
[My taffybar config](https://github.com/benley/dotfiles/blob/d1d4f8f1ccdde8339851adeecc8fbc60c4e35feb/cfg/.config/taffybar/taffybar.hs)

Open a window whose title contains "🇺🇸🗳🇺🇸", or probably any other sequence of color emoji characters.  Twitter in a Chromium tab is a common place for this to happen.

You may need [something like this in ~/.fonts.conf](https://github.com/benley/dotfiles/blob/6cce95299846d43f372fd26dbbdc482458afc6f0/machines/imports/fonts.conf) and some fonts installed.  On NixOS I do that [with this module](https://github.com/benley/dotfiles/blob/6cce95299846d43f372fd26dbbdc482458afc6f0/machines/imports/fonts.nix)

**Expected behavior**
Taffybar should stay the same size and display the window title.

**Screenshots**
Before:
![taffybar-before](https://user-images.githubusercontent.com/1432730/45054146-1a531d80-b05a-11e8-90fb-5eff1382c498.png)

After:
![taffybar-after](https://user-images.githubusercontent.com/1432730/45054158-2343ef00-b05a-11e8-9224-aa7209badf3c.png)

**Version information**
Taffybar version (or git sha if building from source): 09c600cdc1332619087a90c3958d5d1fd68a6067
Installation method (nix/stack/cabal/arch packages): Nix

Gtk version: 3.22.20, from nixos/nixpkgs@8395f9aa85e621c076334a67d814de8221ce7983

**Comment by benley** [comment]
Interesting.  Is there a way for taffybar to at least be aware of its actual height and update the strut height accordingly?  It would be kinda cool if taffybar.hs didn't set a height in pixels at all and that all happened in CSS.

**Comment by IvanMalison** [comment]
> It would be kinda cool if taffybar.hs didn't set a height in pixels at all and that all happened in CSS.

Yes, I have wanted to do things this way for a long time. Unfortunately, the taffybar internals need to know about the height in order to allocate the strut space.

> Is there a way for taffybar to at least be aware of its actual height and update the strut height accordingly?

I think what you are asking here is whether taffybar can dynamically set the strut allocation and window size.

There are a couple of reasons this isn't possible:

- At least in the case of xmonad, once a window is marked as "being a strut" all property updates from that window are ignored.
- Taffybar can't actually force a window resize of itself, because window managers are free to ignore resize requests. While xmonad DOES handle resize requests (I believe) it ignores taffybar's requests because it ignores its window

**Comment by benley** [comment]
Ah, got it.  So in order to change the height of the strut space, taffybar would have to delete and recreate its window?

**Comment by IvanMalison** [comment]
> taffybar would have to delete and recreate its window?

Right. Which is actually quite possible.

Check out https://github.com/travitch/taffybar/blob/09c600cdc1332619087a90c3958d5d1fd68a6067/src/System/Taffybar/DBus/Toggle.hs#L16

if you want to be able to restart taffybar manually.

**Comment by IvanMalison** [comment]
@benley Did you ever get to the bottom of this in any way? Figure out a workaround?

**Comment by evanjs** [comment]
One interesting thing I've found is that this does not seem to trigger on emoji displayed as images, such as those in the [cargo-about](https://github.com/EmbarkStudios/cargo-about) repository.
Of course, repositories such as [sqlx](https://github.com/launchbadge/sqlx) do trigger this behavior.

**Comment by IvanMalison** [comment]
@evanjs That's not really surprising. The emoji seems to only cause an issue when something happens with the font (the emoji size is not the same as the main font size)

**Comment by evanjs** [comment]
@IvanMalison Ahh that makes sense.  So basic Unicode glyphs in general are not the issue, but rather, glyphs that differ in size from the rest of the font. Hrm.
